util: add util/errors

jasnell opened the pull request:
Exports a subset of `internal/errors` as `require('util/errors')`.

The reason for `util/errors` is because there is already a `errors`
module on npm that has a non-trivial number of downloads per week.
Unfortunately that makes it `semver-major`

Signed-off-by: James M Snell <jasnell@gmail.com>
Fixes: https://github.com/nodejs/node/issues/38361
Fixes: https://github.com/nodejs/node/issues/14554

mcollina commented:
I'm +1

mcollina commented:
I think we should also expose the mechanism to create errors-with-codes. I did https://github.com/fastify/fastify-error to map the behavior of core.

jasnell commented:
@mcollina 
> I'm +1

Which are you +1 about? That we *should* export all of the `ERR_` constructors?

mcollina commented:
I'm +1 to exporting the JS ones, e.g. `internal/errors`.

jasnell commented:
@mcollina @ronag ....

Ok, I've added a utility method to this that allows creating custom error types with codes...

```js
const {
  makeErrorWithCode,
} = require('util/errors');

const messages = new Map([
  ['FOO', 'hello %s'],
  ['BAR', (name) => `hello ${name}`],
]);

const FOO = makeErrorWithCode('FOO', { Base: TypeError, messages });

throw new FOO('world');  // message: hello world
```

I've added basic documentation. What I'm not sure about is how we should document the constructors for the various `ERR_` errors.

joyeecheung commented:
> Should we export all of the ERR_* constructors like this?

What's the upside of exporting the constructors? The constructors take custom arguments and without documentations of the signatures they would be rather difficult to use, even though the sigantures can be quite unstable, since we change the message formatters all the time, adding more details. Doing a `instanceof` check with the constructors also seems less reliable than just checking the `error.code` property, which was what these errors were designed for.

mcollina commented on doc/api/errors.md:
```diff
@@ -2756,7 +2769,108 @@ removed: v15.0.0
+## The `util/errors` API
+<!-- YAML
+added: REPLACEME
+-->
```
Should we document `codes`?

jasnell replied:
No, let's not expose them for now.

mcollina reviewed (commented):
Good work!

jasnell commented:
Ok, I've updated this to *remove* the exposed `ERR_` constructors for the expressed by both @targos and @joyeecheung ... I just don't think we're ready to make those constructors public and commit to an API contract for those yet. Everything else remains the same except that the documentation moves to `util.md`

aduh95 commented on lib/util/errors.js:
```diff
@@ -0,0 +1,65 @@
+  },
+} = require('internal/errors');
+
+const {
+  DOMException
+} = internalBinding('messaging');
```
```suggestion
const {
  DOMException,
} = internalBinding('messaging');
```
or
```suggestion
const { DOMException } = internalBinding('messaging');
```

jasnell replied:
I prefer to keep it across multiple lines

aduh95 commented on lib/util/errors.js:
```diff
@@ -0,0 +1,65 @@
+    Base = Error,
+    messages = new SafeMap(),
+  } = options;
+  if (typeof Base !== 'function') {
+    throw new ERR_INVALID_ARG_TYPE(
+      'options.Base',
+      'Function',
+      Base);
+  }
```
```suggestion
  validateFunction(Base, 'options.Base');
```

aduh95 commented on lib/util/errors.js:
```diff
@@ -0,0 +1,65 @@
+      messages);
+  }
+  const message = messages.get(code);
+  if (message === undefined) {
```
We may want to add an extra check in case the code is not missing but mapped to `undefined`. It's the error path anyway.
```suggestion
  if (message === undefined && !messages.has(code)) {
```

jasnell replied:
This is unnecessary given the `message.get(code)` immediately before. `message` will be `undefined` in either case.

aduh95 replied:
> `message` will be `undefined` in either case.

My point exactly, I don’t think we should treat a value of `undefined` the same as a missing value.

```js
const messages = new Map;
messages.set('foo', undefined);
makeErrorWithCode('foo', { messages }); // I expect a ERR_INVALID_ERROR_MESSAGE error.
makeErrorWithCode('bar', { messages }); // I expect a ERR_MISSING_ERROR_MESSAGE error.
```

It would be nice to add a test case for this BTW.

jasnell replied:
> My point exactly, I don’t think we should treat a value of undefined the same as a missing value.

Hmmm... I don't really agree with that. I will add a test tho :-)

mcollina reviewed (approved):
lgtm

jasnell commented:
@nodejs/tsc ... this needs another TSC review and sign off

targos commented on lib/util/errors.js:
```diff
@@ -0,0 +1,61 @@
+  }
+  if (typeof message !== 'function' &&
+      typeof message !== 'string') {
+    throw new ERR_INVALID_ERROR_MESSAGE(code);
```
Why not ERR_INVALID_ARG_TYPE ? It supports multiple types and is used almost everywhere else for this kind of error.

targos replied:
I realize `message` is not an arg here. Why is it only validated now rather than before adding it to the messages map?

targos replied:
Please let me think a bit about the proposed API. There's something that doesn't feel right.

jasnell replied:
The API was designed with minimal changes to the internals in mind.

ronag reviewed (approved):
LGTM

jasnell commented:
Note to folks: While this has the sign off from the two TSC members necessary to land, @targos has requested more time to review it. Please do not land this PR until they've had a few days to complete their review.

mhdawson commented:
@James I'm wondering about the motivation/use case. Could you give us a 1 paragraph level description of the problem being solved, use case for the exposed APIs?

jasnell commented:
@mhdawson ... See https://github.com/fastify/fastify-error for an example use case motivating the api for creating the error classes.

Regarding the use case for exposing AbortError and DOMException, it comes down to consistency. We want users to make use of things like AbortSignal. It makes sense to make it easy for them to do so in a way that is consistent with core's use.

mhdawson commented:
@jasnell thanks, that helps :)

jasnell commented:
@targos ... any further thoughts on this?

targos commented:
Sorry, I didn't find time to look at this during the long weekend. I'll do it first time tomorrow morning!

BridgeAR reviewed (changes requested):
I am not certain that we should really expose our `makeErrorWithCode`, even with a layer above abstracting it.

There's also at least one difference to our API: Our internals allow to define multiple error classes for the same error code. That is not possible with the new API.

If we really want to expose this, we should also document, that input is going to be validated while creating a new instance during runtime. The error being thrown at this point is going to point to Node.js though, that it's an implementation mistake on our side. This would from now on definitely be wrong and we would have to update that to throw something else instead. And if I am not mistaken, it's possible to create errors with identical codes that are already in use by e.g., Node.js. This is not ideal. I would also rather differentiate codes that are used by Node.js and ones that are created by libraries as we might add duplicate codes that way later on.

For these reasons I am currently against adding this to core.

jasnell commented:
> There's also at least one difference to our API: Our internals allow to define multiple error classes for the same error code. That is not possible with the new API.

That was intentional to simplify the API a bit. There's nothing to stop the user code from defining multiple custom errors with separate types and the same code.

> If we really want to expose this, we should also document, that input is going to be validated while creating a 
new instance during runtime. The error being thrown at this point is going to point to Node.js though, that it's an implementation mistake on our side. This would from now on definitely be wrong and we would have to update that to throw something else instead.

That's fine. We can change the error that is thrown when this API is used.

> And if I am not mistaken, it's possible to create errors with identical codes that are already in use by e.g., Node.js. This is not ideal. I would also rather differentiate codes that are used by Node.js and ones that are created by libraries as we might add duplicate codes that way later on.

I'm not convinced that's really a problem.

jasnell commented:
Removed the `author ready` label while we get the objection resolved.

DerekNonGeneric reviewed (changes requested):
Agreed that the `code` property is going to be subject to debate here.

Thanks for putting a hold on this PR for the time being. AFAICT, it does still seem to be half-baked. 
 
Changing it up so that the error codes give us a way to know if the error came from Node.js internals or not seems smart.

jasnell commented:
@DerekNonGeneric @BridgeAR ... Ok, but I'm going to need a bit more detail in terms of what *would* be acceptable to you both here. It's clear that you don't want what the API currently does but it's not clear what you would want.

mcollina commented:
I would like to focus on the _need_ for exposing an API to create errors with code, and leave out the current implementation to a later discussion.

Long time ago, we underwent a massive refactoring to add a `code` property to all Node.js errors. We did this change because developers relied on the `.message` property of errors, which lead to high unstable code. I believe this was an extremely useful addition to Node.js. I think we can all agree to the benefits of this, and if not please voice your concerns clearly.

As we have clarified that creating errors with a `.code` property is useful, we are limiting our users by not exposing a utility to do so themselves. They will need to implement their own system, if at all. I think we can promote a _better pattern_ for the whole ecosystem by recommending them to add `.code` to their errors by exposing such an utility.
As an example, check out https://github.com/fastify/fastify-error. 

The point of debate is if we agree something like this is worthwhile doing first and then figure out the best implementation and what to expose.

mcollina commented:
I can coordinate moving https://github.com/fastify/fastify-error if we want to make it official.

My assumption is that we would link the module from the Node.js docs?

mhdawson commented:
If we moved it over I think referencing in the docs would make sense to me.

jasnell commented:
Going to go ahead and close this PR given that it likely won't be moving forward as is.

mhdawson commented:
@jasnell  In https://github.com/nodejs/node/pull/38575#issuecomment-849121485 I was just asking about different options but if moving into org satisfies the need from your perspective then seems like a reasonable way to move forward.
